test-data/unit/check-classes.test

```diff
@@ -894,6 +892,281 @@ a.f = a.f # E: Property "f" defined in "A" is read-only
 a.f.x # E: "int" has no attribute "x"
 [builtins fixtures/property.pyi]
 
+-- Descriptors
+-- -----------
+
+
+[case testAccessingNonDataDescriptor]
+class D():
+    def __get__(self, instance: Any, owner: Any) -> str:
+        return 'some string'
+
+class A:
+    f = D()
+
+a = A() # type: A
+reveal_type(a.f)  # E: Revealed type is 'builtins.str'
+
+[case testSettingDataDescriptor]
+class D():
+    def __get__(self, instance: Any, owner: Any) -> str:
+        return 'some string'
+
+    def __set__(self, instance: Any, value: str) -> None:
+        pass
+
+class A:
+    f = D()
+
+a = A() # type: A
+a.f = ''
+a.f = 1 # E: Incompatible types in assignment (expression has type "int", variable has type "str")
+
+[case testReadingDescriptorWithoutDunderGet]
+from typing import Union
+
+class D():
+    def __set__(self, instance: Any, value: str) -> None:
+        pass
+
+class A:
+    f = D()  # type: Union[D, str]
+
+    def __init__(self):
+        self.f = 'some string'  # type: str
+
+a = A() # type: A
+reveal_type(a.f)  # E: Revealed type is 'Union[__main__.D, builtins.str]'
+
+[case testAccessingGenericNonDataDescriptor]
+from typing import TypeVar, Type, Generic
+
+T = TypeVar('T')
+V = TypeVar('V')
+class D(Generic[T, V]):
+    def __init__(self, value: V) -> None:
+        self.value = value
+    def __get__(self, instance: T, owner: Type[T]) -> V:
+        return self.value
+
+class A:
+    f = D(10)  # type: D[A, int]
+    g = D('10')  # type: D[A, str]
+
+a = A() # type: A
+reveal_type(a.f)  # E: Revealed type is 'builtins.int*'
+reveal_type(a.g)  # E: Revealed type is 'builtins.str*'
+
+[case testSettingGenericDataDescriptor]
+from typing import TypeVar, Type, Generic
+
+T = TypeVar('T')
+V = TypeVar('V')
+class D(Generic[T, V]):
+    def __init__(self, value: V) -> None:
+        self.value = value
+    def __get__(self, instance: T, owner: Type[T]) -> V:
+        return self.value
+    def __set__(self, instance: T, value: V) -> None:
+        pass
+
+class A:
+    f = D(10)  # type: D[A, int]
```

Comment:
When I run the tests without these annotations, I get "Need type annotation for variable", pointing to the line where the descriptor is defined.

Testing with other annotations, I see that it's an issue with the generic `T`, in particular. I fixed up the tests to remove `T` from the class-level generics, which allowed me to remove a bunch of the annotations that you thought should be unnecessary.